### PR TITLE
[Inference/Spec-Dec] Add Speculative Decoding and GLIDE Spec-Dec

### PR DESCRIPTION
**Pull request by yuanheng-zhao**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#5245 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

Implement and add feature: Speculative Decoding into ColossalAI-Inference.
- Revision on Triton Kernels
- Drafter Model Container
- Support for Speculative Decoding
- Support for GLIDE Model (https://arxiv.org/pdf/2402.02082.pdf) 

Future Work:
1. TP support on the main model (optional)
2. `q_len > 1` support on CUDA kernel, used by the main model during verification
3. Optimize Spec-Dec step process in InferenceEngine

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by yuanheng-zhao** [comment]
image

**Review by yuehuayingxueluo — APPROVED** [review]

File: colossalai/kernel/triton/flash_decoding.py
Comment (NonvolatileMemory) [inline colossalai/kernel/triton/flash_decoding.py:1]: do we use falsh decoding in the verification branch?

I think if the kernel is better to be aligned with prefill stage, if the speculation lengths are not very small.

**Review by NonvolatileMemory — COMMENTED** [review]
🐮

File: colossalai/inference/spec/drafter.py
[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply colossalai/inference/spec/drafter.py:1]: `trim_kv_cache` is used for trimming invalid kv cache of the small model (drafter model)

[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply colossalai/inference/spec/drafter.py:1]: will fix it

File: colossalai/inference/spec/README.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+# use the same tokenizer for both the main model and the drafter model
+tokenizer = AutoTokenizer.from_pretrained(model_path_or_name)
+tokenizer.pad_token = tokenizer.eos_token
+
+# drafter model
+drafter_model_path_or_name = "REPLACE_TO_LLAMA_68M_PATH_OR_MODEL_CARD"
+drafter_model = AutoModelForCausalLM.from_pretrained(drafter_model_path_or_name)
+
+# Initialize the inference engine
+inference_config = InferenceConfig(
```
[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply colossalai/inference/spec/README.md:41]: I'll create another PR for fixing docstrings and comment, as the current one is a bundle of several PRs which have been merged into the feature branch. Will rebase and merge these commits to `feature/colossal-infer` first

File: colossalai/inference/spec/drafter.py
[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply colossalai/inference/spec/drafter.py:1]: Comments/Docstrings will be fixed in another PR.

**Comment by yuanheng-zhao** [comment]
Inference CI issue (i.e. dependency and example test) to be fixed in another PR.

**Pull request merged**